presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,346 @@
+            return Optional.empty();
+        }
+
+        MultiJoinNode multiJoinNode = toMultiJoinNode(joinNode, lookup);
+        return Optional.of(new JoinEnumerator(idAllocator, symbolAllocator, session, lookup, multiJoinNode.getFilter(), costComparator).chooseJoinOrder(multiJoinNode.getSources(), multiJoinNode.getOutputSymbols()));
+    }
+
+    @VisibleForTesting
+    public static class JoinEnumerator
+    {
+        private final Map<Set<PlanNode>, PlanNode> memo = new HashMap<>();
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final Ordering<PlanNode> planNodeOrdering;
+        private final EqualityInference allInference;
+        private final Expression allFilter;
+
+        public JoinEnumerator(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session, Lookup lookup, Expression filter, CostComparator costComparator)
+        {
+            requireNonNull(idAllocator, "idAllocator is null");
+            requireNonNull(symbolAllocator, "symbolAllocator is null");
+            requireNonNull(session, "session is null");
+            requireNonNull(lookup, "lookup is null");
+            requireNonNull(filter, "filter is null");
+            requireNonNull(costComparator, "costComparator is null");
+            this.idAllocator = idAllocator;
+            this.session = session;
+            this.planNodeOrdering = getPlanNodeOrdering(costComparator, lookup, session, symbolAllocator);
+            this.allInference = createEqualityInference(filter);
+            this.allFilter = filter;
+        }
+
+        private static Ordering<PlanNode> getPlanNodeOrdering(CostComparator costComparator, Lookup lookup, Session session, SymbolAllocator symbolAllocator)
+        {
+            return new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+        }
+
+        private PlanNode chooseJoinOrder(List<PlanNode> sources, List<Symbol> outputSymbols)
+        {
+            Set<PlanNode> multiJoinKey = ImmutableSet.copyOf(sources);
+            PlanNode node = memo.get(multiJoinKey);
+            if (node == null) {
+                checkState(sources.size() > 1);
+                node = generatePartitions(sources.size())
+                        .map(partitioning -> createJoinAccordingToPartitioning(sources, outputSymbols, partitioning))
+                        .filter(Optional::isPresent)
+                        .map(Optional::get)
+                        .min(planNodeOrdering)
```

Comment:
It may happen that some sources (source tables) do not have statistics necessary to compute cost. Even if `CostComparator` sorted unknown values as e.g. last, it wouldn't work for `ReorderJoins`, as RJ would then choose "random" (first?) join order encountered. Thus, unknown costs must be handled explicitly here. I'd opt for backing out and returning `Optional.empty()` from the rule if any encountered cost was unknown.

To avoid giving wrong advices, `CostComparator` will throw when any of the compared costs is unknown. 

cc @losipiuk